### PR TITLE
Dashboard

### PR DESCRIPTION
**Pull request by axelbacklund**
Starting the creation of dashboard. The protection views cannot be expanded yet and strings are placeholders.

### CODE REVIEW

File: Src/Screens/DashboardScreen/Dashboard.swift [hunk 1]
```diff
@@ -33,10 +33,12 @@ extension Dashboard: Presentable {
             String(.DASHBOARD_BANNER_ACTIVE_TITLE(firstName: $0))
         }.bindTo(viewController, \.navigationItem.title)
 
-        let view = UIView()
-        view.backgroundColor = .purple
-
-        viewController.view = view
+        let form = FormView()
+        
+        let myProtectionSection = DashboardSection()
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Dashboard.swift]: `DashboardSection` or `myProtection` 🤔

File: Src/Screens/DashboardScreen/Viewables/ProtectionView.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        protections: [String] = []
+    ) {
+        self.titleText = title
+        self.iconAsset = icon
+        self.color = color
+        self.protections = protections
+    }
+}
+
+extension ProtectionView: Viewable {
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ProtectionView.swift]: As discussed IRL find a more appropriate name for this

File: Src/Screens/DashboardScreen/Viewables/ExpandableProtectionRow.swift [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    init(
+        content: Content,
+        expandableContent: ExpandableContent
+    ) {
+        self.contentViewable = content
+        self.expandableContentViewable = expandableContent
+    }
+}
+
+extension ExpandableProtectionRow: Viewable {
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableProtectionRow.swift]: maybe just `ExpandableRow` now when you made this generic? :)

File: Src/Screens/DashboardScreen/Viewables/ExpandableProtectionRow.swift [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+//
+//  Created by Axel Backlund on 2019-03-29.
+//
+
+import Flow
+import Form
+import Foundation
+import UIKit
+
+struct ExpandableProtectionRow<Content: Viewable, ExpandableContent: Viewable> where
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableProtectionRow.swift]: there must be a better way to do this  🤔 but it's fine for now.

File: Src/Screens/DashboardScreen/Viewables/ExpandableProtectionRow.swift [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    Content.Matter == UIView,
+    Content.Events == ViewableEvents,
+    Content.Result == Disposable,
+    ExpandableContent.Matter == UIView,
+    ExpandableContent.Events == ViewableEvents,
+    ExpandableContent.Result == Disposable
+    {
+    
+    let contentViewable: Content
+    let expandableContentViewable: ExpandableContent
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableProtectionRow.swift]: name this `expandedContent` maybe?

File: Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        let titleLabel = MultilineLabel(styledText: StyledText(text: titleText, style: .rowTitle))
+        bag += titlesView.addArranged(titleLabel)
+        
+        let subtitleLabel = MultilineLabel(styledText: StyledText(text: subtitleText, style: .rowSubtitle))
+        bag += titlesView.addArranged(subtitleLabel)
+        
+        containerStackView.addArrangedSubview(titlesView)
+        
+        // Chevron down
+        let chevronDown = Icon(frame: .zero, icon: Asset.chevronRight, iconWidth: 30)
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift]: I think you can omit `frame: .zero` it should default to that.

File: Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        chevronDown.transform = CGAffineTransform.init(rotationAngle: CGFloat.pi / 2)
+        chevronDown.setContentHuggingPriority(.defaultHigh, for: .horizontal)
+        containerStackView.addArrangedSubview(chevronDown)
+        
+        bag += containerStackView.makeConstraints(wasAdded: events.wasAdded).onValue({ make, _ in
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift]: you can remove the parenthesis on the `onValue`.

File: Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift [hunk 8]
```diff
@@ -56,31 +48,60 @@ extension LargeIconTitleSubtitle: Viewable {
         containerStackView.isLayoutMarginsRelativeArrangement = true
         
         // Large icon
-        let icon = Icon(icon: iconAsset, iconWidth: iconWidth)
+        let icon = Icon(icon: Asset.homePlain, iconWidth: iconWidth)
         icon.setContentHuggingPriority(.defaultHigh, for: .horizontal)
         containerStackView.addArrangedSubview(icon)
         
+        bag += imageSignal.atOnce().filter { $0 != nil }.map { $0! }.bindTo(icon, \.icon)
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift]: you could simply do `compactMap { $0 }` and remove both the filter and map.

File: Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+//  Created by Axel Backlund on 2019-03-29.
+//
+
+import Flow
+import Form
+import Foundation
+import UIKit
+
+struct ExpandableRow<Content: Viewable, ExpandableContent: Viewable> where
+    Content.Matter == UIView,
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift]: This should probably be a Generic that extends UIView, so:

```suggestion
    Content.Matter == ContentMatter,
```

File: Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+//
+//  Created by Axel Backlund on 2019-03-29.
+//
+
+import Flow
+import Form
+import Foundation
+import UIKit
+
+struct ExpandableRow<Content: Viewable, ExpandableContent: Viewable> where
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift]:
```suggestion
struct ExpandableRow<Content: Viewable, ContentMatter: UIView, ExpandableContent: Viewable> where
```

**Comment by axelbacklund** [comment]
Done with all functionality and translations!

File: Src/Generic/Animations/CoreSignal+Animation.swift [hunk 11]
```diff
@@ -10,9 +10,10 @@ import Flow
 import Foundation
 import UIKit
 
-struct AnimatedSignal<Value>: SignalProvider, Disposable {
+struct AnimatedSignal<Value, From: SignalProvider>: SignalProvider, Disposable {
```
Comment (Iamsamwhoami) [inline Src/Generic/Animations/CoreSignal+Animation.swift]: `From` can be removed.

File: Src/Generic/Viewables/CheckmarkLabel.swift [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+struct CheckmarkLabel {
+    let styledTextSignal: ReadWriteSignal<StyledText>
+    
+    init(styledText: StyledText) {
+        self.styledTextSignal = ReadWriteSignal(styledText)
+    }
+}
+
+extension CheckmarkLabel: Viewable {
```
Comment (Iamsamwhoami) [inline Src/Generic/Viewables/CheckmarkLabel.swift]: Does this have to be so specific, couldn't it be just a label with prepending icon?

File: Src/Generic/Viewables/Divider.swift [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+//
+
+import Flow
+import Foundation
+import UIKit
+
+struct Divider {
+    let backgroundColor: UIColor
+    
+    init(backgroundColor: UIColor) {
```
Comment (Iamsamwhoami) [inline Src/Generic/Viewables/Divider.swift]: This initializer is unnecessary.

File: Src/Screens/DashboardScreen/Viewables/ChatActionsSection.swift [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        let bag = DisposeBag()
+        
+        let stackView = UIStackView()
+        stackView.axis = .vertical
+        stackView.spacing = 16
+        
+        let headerLabelContainer = UIStackView()
+        headerLabelContainer.edgeInsets = UIEdgeInsets(top: 0, left: 16, bottom: 0, right: 16)
+        let headerLabel = MultilineLabel(styledText: StyledText(
+            text: String(key: .DASHBOARD_CHAT_ACTIONS_HEADER),
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ChatActionsSection.swift]: NIT: Indentation is off here, should be tabbed another level.

File: Src/Screens/DashboardScreen/Viewables/ChatActionsSection.swift [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+        bag += dataSignal.atOnce()
+            .compactMap { $0?.filter { $0?.enabled == true } }
+            .onValue { chatActions in
+                
+            buttonStackView.subviews.forEach { view in
+                view.removeFromSuperview()
+            }
+                
+            for chatAction in chatActions  {
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ChatActionsSection.swift]: NIT: extra space after chatActions

File: Src/Screens/DashboardScreen/Viewables/ChatActionsSection.swift [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+                let buttonContainer = UIView()
+                let button = Button(title: chatAction?.text ?? "", type: .standard(backgroundColor: .purple, textColor: .white))
+                bag += button.onTapSignal.onValue {
+                    DashboardRouting.openChat(viewController: self.presentingViewController, chatActionUrl: chatAction?.triggerUrl ?? "")
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ChatActionsSection.swift]: should we even open chat if we don't have a triggerUrl?

File: Src/Screens/DashboardScreen/Viewables/CountdownShapes.swift [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+            make.height.equalTo(12)
+        }
+        barStack.addArrangedSubview(secondBar)
+        
+        view.addArrangedSubview(barStack)
+        
+        view.snp.makeConstraints { (make) in
+            make.height.centerX.centerY.equalToSuperview()
+        }
+
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/CountdownShapes.swift]: NIT: Double new lines

File: Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+            containerView.layer.cornerRadius = 15
+            containerView.layer.shadowOpacity = 0.15
+            containerView.layer.shadowOffset = CGSize(width: 0, height: 6)
+            containerView.layer.shadowRadius = 8
+            containerView.layer.shadowColor = UIColor.darkGray.cgColor
+        } else {
+            containerView.backgroundColor = .transparent
+        }
+        
+        //containerView.clipsToBounds = true
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift]: Remove this

File: Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+        //containerView.layer.masksToBounds = true
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift]: Remove this

File: Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+                let delay: Double = isOpen ? 0 : 0.15
+                let opacity: CGFloat = isOpen ? 1 : 0
+                
+                bag += Signal(after: delay).animated(style: AnimationStyle.easeOut(duration: 0.1), animations: { _ in
+                    dividerView.alpha = opacity
+                })
+            })
+        }
+        
+        // TODO: Rewrite the above to a nice Signal thing like below. Currently doesn't work.
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/ExpandableRow.swift]: TODO here

File: Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+        
+        let subtitleLabel = MultilineLabel(styledText: StyledText(text: "", style: .rowSubtitle))
+        bag += titlesView.addArranged(subtitleLabel)
+        
+        bag += subtitleSignal.atOnce()
+            .map { StyledText(text: $0, style: .rowSubtitle) }
+            .bindTo(subtitleLabel.styledTextSignal)
+        
+        containerStackView.addArrangedSubview(titlesView)
+
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift]: NIT: Multiple new lines

File: Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+        
+        containerStackView.addArrangedSubview(titlesView)
+        
+        
+        let chevronDown = Icon(icon: Asset.chevronRight, iconWidth: 25)
+        chevronDown.setContentHuggingPriority(.defaultHigh, for: .horizontal)
+        containerStackView.addArrangedSubview(chevronDown)
+        
+        func setChevronRotation(isOpen: Bool, animate: Bool) {
+            let rotationAngle = isOpen ? ((3 * CGFloat.pi / 2) * 1.0001) : (CGFloat.pi / 2)
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/LargeIconTitleSubtitle.swift]: hmm, the `1.0001` feels like a hack 🤔

File: Src/Screens/DashboardScreen/Viewables/PaymentNeedsSetupSection.swift [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+        bag += infoContainer.add(infoLabel) { labelView in
+            labelView.textAlignment = .center
+            labelView.snp.makeConstraints { make in
+                make.height.width.centerY.centerX.equalToSuperview()
+            }
+        }
+        containerStackView.addArrangedSubview(infoContainer)
+        
+        let buttonContainer = UIView()
+        let connectButton = Button(title: String(key: .DASHBOARD_PAYMENT_SETUP_BUTTON), type: .outline(borderColor: .purple, textColor: .purple))
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PaymentNeedsSetupSection.swift]: Wrap this line

File: Src/Screens/DashboardScreen/Viewables/PendingInsurance.swift [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+        stackView.edgeInsets = UIEdgeInsets(top: 15, left: 25, bottom: 30, right: 25)
+        
+        bag += dataSignal.atOnce().map { $0 == nil || $0?.status != .inactive || $0?.status != .inactiveWithStartDate }.bindTo(stackView, \.isHidden)
+        
+        bag += dataSignal.atOnce().compactMap { $0 } .onValue { insurance in
+            switch insurance.status {
+            case .inactive:
+                let content = CountdownShapes()
+                let moreInfo = PendingInsuranceMoreInfo()
+                let expandableView: ExpandableRow<CountdownShapes, PendingInsuranceMoreInfo> = ExpandableRow(content: content, expandedContent: moreInfo, transparent: true)
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PendingInsurance.swift]: Shouldn't the type be inferred here?

File: Src/Screens/DashboardScreen/Viewables/PendingInsurance.swift [hunk 25]
```diff
@@ -0,0 +1,102 @@
...
+                    print("pls update ios")
+                }
+                ()
+            default:
+                ()
+            }
+        }
+        
+        func addBottomContent<G: Viewable>(view: ExpandableRow<G, PendingInsuranceMoreInfo>) {
+
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PendingInsurance.swift]: NIT: Starting newline in func

File: Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+        view.edgeInsets = UIEdgeInsets(top: 20, left: 0, bottom: 10, right: 0)
+        
+        view.snp.makeConstraints { make in
+            make.width.lessThanOrEqualTo(400)
+        }
+        
+        var text = ""
+        
+        if let _date = date {
+            // - TODO: Date in app's language
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift]: Use current locale.

File: Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+        
+        let view = UIStackView()
+        view.axis = .vertical
+        view.edgeInsets = UIEdgeInsets(top: 20, left: 0, bottom: 10, right: 0)
+        
+        view.snp.makeConstraints { make in
+            make.width.lessThanOrEqualTo(400)
+        }
+        
+        var text = ""
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift]: you could simply do:
```
var text: String {
 // do your computation here
}
```

Instead.

File: Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+            dateFormatter.dateStyle = .long
+            let dateString = dateFormatter.string(from: _date)
+            
+            text = String(key: .DASHBOARD_PENDING_HAS_DATE(date: dateString))
+            
+        } else {
+            text = String(key: .DASHBOARD_PENDING_NO_DATE)
+        }
+        
+        //let text = date != nil ? String(key: .DASHBOARD_PENDING_HAS_DATE(date: "31 maj 2019")) : String(key: .DASHBOARD_PENDING_NO_DATE)
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift]: Remove this.

File: Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+        view.axis = .vertical
+        view.edgeInsets = UIEdgeInsets(top: 20, left: 0, bottom: 10, right: 0)
+        
+        view.snp.makeConstraints { make in
+            make.width.lessThanOrEqualTo(400)
+        }
+        
+        var text = ""
+        
+        if let _date = date {
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/PendingInsuranceMoreInfo.swift]: why is the variable named `_date`?

File: Src/Screens/DashboardScreen/Viewables/Peril.swift [hunk 30]
```diff
@@ -0,0 +1,124 @@
...
+    
+    init(title: String, id: String, description: String, presentingViewController: UIViewController) {
+        self.title = title
+        self.id = id
+        self.description = description
+        self.presentingViewController = presentingViewController
+    }
+}
+
+// query chatactions
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/Peril.swift]: Remove this

File: Src/Screens/DashboardScreen/Viewables/Peril.swift [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+
+extension Peril: Reusable {
+
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/Peril.swift]: NIT: newline

File: Src/Screens/DashboardScreen/Viewables/Peril.swift [hunk 32]
```diff
@@ -0,0 +1,124 @@
...
+                return Asset.stuffFire
+            case "STUFF.BRF.WATER", "STUFF.RENT.WATER", "STUFF.SUBLET.BRF.WATER", "STUFF.SUBLET.RENT.WATER":
+                return Asset.stuffWaterLeak
+            case "STUFF.BRF.WEATHER", "STUFF.RENT.WEATHER", "STUFF.SUBLET.BRF.WEATHER", "STUFF.SUBLET.RENT.WEATHER":
+                return Asset.stuffWeather
+            default:
+                return Asset.houseBreakIn
+            }
+        }
+
```
Comment (Iamsamwhoami) [inline Src/Screens/DashboardScreen/Viewables/Peril.swift]: NIT: Newlines

File: Src/Screens/DashboardScreen/Viewables/Peril.swift [hunk 33]
```diff
@@ -0,0 +1,119 @@
...
+                }
+                
+                titleLabel.sizeToFit()
+            }
+            
+            let tapGesture = UITapGestureRecognizer()
+            bag += perilView.install(tapGesture)
+            
+            bag += tapGesture.signal(forState: .ended).onValue { _ in
+                let title = peril.title.replacingOccurrences(of: "-\n", with: "")
```
[the remark this answers is not in the document]
Comment (axelbacklund) [reply Src/Screens/DashboardScreen/Viewables/Peril.swift]: Will fix when adding translations to perils!

**Pull request merged**